task_processing/plugins/mesos/logging_executor.py

```diff
@@ -0,0 +1,208 @@
+import time
+from threading import Lock
+from threading import Thread
+from urllib.parse import urlparse
+
+import requests
+from pyrsistent import field
+from pyrsistent import m
+from pyrsistent import PMap
+from pyrsistent import pmap
+from pyrsistent import PRecord
+from pyrsistent import v
+from six.moves.queue import Queue
+
+from task_processing.interfaces.task_executor import TaskExecutor
+
+
+log = logging.getLogger(__name__)
+logging.getLogger("urllib3").setLevel(logging.WARNING)
+
+# Read task log in 4K chunks
+TASK_LOG_CHUNK_LEN = 4096
+
+
+class LogMetadata(PRecord):
+    log_url = field(type=str, initial='')
+    log_path = field(type=str, initial='')
+    log_offsets = field(type=PMap,
+                        factory=pmap,
+                        initial=pmap({'stdout': 0, 'stderr': 0}))
+    container_id = field(type=str, initial='')
+    executor_id = field(type=str, initial='')
+
+
+class MesosLoggingExecutor(TaskExecutor):
+    def __init__(self, downstream_executor):
+        self.downstream_executor = downstream_executor
+
+        self.src_queue = downstream_executor.get_event_queue()
+        self.dest_queue = Queue()
+        self.stopping = False
+
+        self.staging_tasks = m()
+        self.running_tasks = m()
+        self.done_tasks = v()
+
+        # A lock is needed to synchronize logging and event processing
+        self.task_lock = Lock()
+
+        self.event_thread = Thread(target=self.event_loop)
+        self.event_thread.daemon = True
+        self.event_thread.start()
+
+        self.logging_thread = Thread(target=self.logging_loop)
+        self.logging_thread.daemon = True
+        self.logging_thread.start()
+
+    def get_task_log_path(self, task_id):
+        log_md = self.running_tasks[task_id]
+        try:
+            response = requests.get(log_md.log_url + '/files/debug').json()
+        except Exception as e:
+            log.error("Failed to fetch files {error}".format(error=e))
+            return
+
+        for key in response.keys():
+            if log_md.executor_id in key and log_md.container_id in key:
+                with self.task_lock:
+                    self.running_tasks = self.running_tasks.set(
+                        task_id,
+                        log_md.set(log_path=key),
+                    )
+                break
+
+    def stream_task_log(self, task_id):
+        if self.running_tasks[task_id].log_path == "":
+            self.get_task_log_path(task_id)
+
+        # Abort in case the log path discovery was not successful
+        log_md = self.running_tasks[task_id]
+        if log_md.log_path == "":
+            return
+
+        offsets = {
+            'stdout': log_md.log_offsets['stdout'],
+            'stderr': log_md.log_offsets['stderr']
+        }
+        agent = urlparse(log_md.log_url).hostname
+
+        for f in ['stdout', 'stderr']:
+            offset = offsets[f]
+            log_path = log_md.log_path + "/" + f
+            while True:
```

Comment:
it could happen that stderr will only be printed after stdout and users will have to order messages, which is not great. but we can fix it later when it actually becomes problem people complain about.